lib/rubygems/package.rb

```diff
@@ -447,12 +447,12 @@ def extract_tar_gz(io, destination_dir, pattern = "*") # :nodoc:
           end
 
         unless directories.include?(mkdir)
-          FileUtils.mkdir_p mkdir, **mkdir_options
+          FileUtils.mkdir_p mkdir, mode: dir_mode ? 0o755 : (entry.header.mode if entry.directory?)
           directories << mkdir
         end
 
         if entry.file?
-          File.open(destination, "wb") {|out| out.write entry.read }
+          File.open(destination, "wb") {|out| IO.copy_stream entry, out }
```

Comment:
Yep, truffleruby failing in your tests too. I haven't been able to extract a distilled case yet, but @eregon could maybe take a look anyway? I think the issue is that copy_stream is not reading and writing byte for byte identically. 

There's layers and layers here, the src entry is a TarReader wrapping a zlib reader, wrapping a TarReader. This change alone causes the bug, if I'm not mistaken.

You may have to revert this part until it's fixed in truffleruby, or found not to be broken?

Reply:
It seems likely to be https://github.com/oracle/truffleruby/issues/3280 which is now fixed in dev builds

Reply:
@martinemde thoughts on introducing a utility method that is `IO.copy_stream` on all rubies except truffleruby before 23.1.2? Could go into https://github.com/rubygems/rubygems/pull/6958

Reply:
That's a great idea. Even fixed it's still broken on older trufflerubies, even if people don't usually update rubygems separately.

Reply:
I can add it to #6958.